Lib/test/test_typing.py

```diff
@@ -447,6 +448,321 @@ def test_bad_var_substitution(self):
+            ('tuple[*Ts]',                             '[*tuple_type[int, ...], *tuple_type[str, ...]]', 'TypeError'),  # Should be tuple[*tuple_type[int, ...], *tuple_type[str, ...]]
+            ('Tuple[*Ts]',                             '[*tuple_type[int, ...], *tuple_type[str, ...]]', 'Tuple[*tuple_type[int, ...], *tuple_type[str, ...]]'),
+
+            ('C[*Ts]',                                 '[*Ts]',                                          'C[*Ts]'),
+            ('tuple[*Ts]',                             '[*Ts]',                                          'tuple[(*Ts,),]'),  # Should be tuple[*Ts]
+            ('Tuple[*Ts]',                             '[*Ts]',                                          'Tuple[*Ts]'),
+
+            ('C[*Ts]',                                 '[T, *Ts]',                                       'C[T, *Ts]'),
+            ('tuple[*Ts]',                             '[T, *Ts]',                                       'TypeError'),  # Should be tuple[T, *Ts]
+            ('Tuple[*Ts]',                             '[T, *Ts]',                                       'Tuple[T, *Ts]'),
+
+            ('C[*Ts]',                                 '[*Ts, T]',                                       'C[*Ts, T]'),
+            ('tuple[*Ts]',                             '[*Ts, T]',                                       'TypeError'),  # Should be tuple[*Ts, T]
+            ('Tuple[*Ts]',                             '[*Ts, T]',                                       'Tuple[*Ts, T]'),
+
+            ('C[T, *Ts]',                              '[int]',                                          'C[int]'),
+            ('tuple[T, *Ts]',                          '[int]',                                          'TypeError'),  # Should be tuple[int]
+            ('Tuple[T, *Ts]',                          '[int]',                                          'Tuple[int]'),
+
+            ('C[T, *Ts]',                              '[int, str]',                                     'C[int, str]'),
+            ('tuple[T, *Ts]',                          '[int, str]',                                     'tuple[int, (str,)]'),  # Should be tuple[int, str]
+            ('Tuple[T, *Ts]',                          '[int, str]',                                     'Tuple[int, str]'),
+
+            ('C[T, *Ts]',                              '[int, str, bool]',                               'C[int, str, bool]'),
+            ('tuple[T, *Ts]',                          '[int, str, bool]',                               'TypeError'),  # Should be tuple[int, str, bool]
+            ('Tuple[T, *Ts]',                          '[int, str, bool]',                               'Tuple[int, str, bool]'),
+
+            ('C[T, *Ts]',                              '[*tuple_type[int, ...]]',                        'C[*tuple_type[int, ...]]'),  # Should be C[int, *tuple_type[int, ...]]
```

Comment:
I think this is a case I updated since your review. The vibe I'm getting from @JelleZijlstra is that the kind of logic complexity he'd be happy with (correct me if I'm wrong, @JelleZijlstra) wouldn't be up to dealing with `C[T, *Ts][*tuple[int, ...]]` - either figuring out the 'right' answer of `C[int, *tuple[int, ...]]`, or accepting it as valid and giving `C[*tuple[int, ...]]`.

Reply:
In fact, according to the updated tentative spec at https://github.com/python/cpython/issues/91162, we'd disallow this altogether - we wouldn't allow any cases where `*tuple[int, ...]` has to be split among multiple things.